cachi2/core/package_managers/metayarn.py

```diff
@@ -0,0 +1,123 @@
+import logging
+import re
+from collections import defaultdict
+from functools import partial
+from pathlib import Path
+from typing import Iterable, Optional, Union
+
+from cachi2.core.errors import PackageRejected
+from cachi2.core.models.input import PackageInput, Request
+from cachi2.core.models.output import RequestOutput
+from cachi2.core.package_managers import yarn, yarn_classic
+from cachi2.core.resolver import _merge_outputs
+from cachi2.core.rooted_path import RootedPath
+
+log = logging.getLogger(__name__)
+
+
+def _contains_yarn_version_trait(
+    package: PackageInput,
+    source_dir: RootedPath,
+    trait_pattern: re.Pattern[str],
+) -> Optional[re.Match[str]]:
+    text = _get_path_to_yarn_lock(package, source_dir).read_text()
+    return trait_pattern.search(text)
+
+
+_yarnv1_pattern = re.compile("yarn lockfile v1")  # See [yarn_v1_trait].
+_yarnv2_pattern = re.compile("__metadata:")  # See [yarn_v2_trait] and [yarn_v2_test_repo].
+contains_yarn_classic = partial(_contains_yarn_version_trait, trait_pattern=_yarnv1_pattern)
+
+
+_yarn_versions = {
+    "yarn_classic": contains_yarn_classic,
+    "yarn": partial(_contains_yarn_version_trait, trait_pattern=_yarnv2_pattern),
+}
```

Comment:
Firstly, :taco: for `partial()` - loving it!

Then:
Zen of Python: "Explicit is better than implicit"
English copyediting class (😆): paraphrasing - "Parallelism is desirable" (i.e. things which serve similar purposes should look similar) 

```python
_yarnv1_pattern = re.compile("yarn lockfile v1")  # See [yarn_v1_trait].
_yarnv2_pattern = re.compile("__metadata:")  # See [yarn_v2_trait] and [yarn_v2_test_repo].
contains_yarn_classic = partial(_contains_yarn_version_trait, trait_pattern=_yarnv1_pattern)
contains_yarnberry = partial(_contains_yarn_version_trait, trait_pattern=_yarnv2_pattern)

_yarn_versions = {
    "yarn_classic": contains_yarn_classic,
    "yarn": contains_yarnberry,
}
```

N.B "yarnberry" is to make it painfully obvious, but the names aren't as important as the structure.

Reply:
Good point, thank you, done. I added `-berry` suffix here, but maybe we should add it everywhere then? (I am also squirreling away "parallelism is desirable", it is a great way of referring to the concept)